diff --git a/clang/lib/DPCT/AnalysisInfo.cpp b/clang/lib/DPCT/AnalysisInfo.cpp
index 9639ea44278a..383a4dea10db 100644
--- a/clang/lib/DPCT/AnalysisInfo.cpp
+++ b/clang/lib/DPCT/AnalysisInfo.cpp
@@ -2564,7 +2564,6 @@ void DpctNameGenerator::printName(const FunctionDecl *FD,
 }
 DpctNameGenerator::DpctNameGenerator(ASTContext &Ctx)
     : G(Ctx), PP(Ctx.getPrintingPolicy()) {
-  PP.PrintCanonicalTypes = true;
 }
 std::string DpctNameGenerator::getName(const FunctionDecl *D) {
   std::string Result;
diff --git a/clang/lib/DPCT/DPCT.cpp b/clang/lib/DPCT/DPCT.cpp
index efad8a5917d6..b842d940d60a 100644
--- a/clang/lib/DPCT/DPCT.cpp
+++ b/clang/lib/DPCT/DPCT.cpp
@@ -184,7 +184,8 @@ UnifiedPath getCudaInstallPath(int argc, const char **argv) {
       Opts.ParseArgs(Argv, MissingArgIndex, MissingArgCount);
 
   // Create minimalist CudaInstallationDetector and return the InstallPath.
-  DiagnosticsEngine E(nullptr, nullptr, nullptr, false);
+  auto DiagOpts = std::make_shared<DiagnosticOptions>();
+  DiagnosticsEngine E(nullptr, *DiagOpts, nullptr, false);
   driver::Driver Driver("", llvm::sys::getDefaultTargetTriple(), E);
   driver::CudaInstallationDetector CudaIncludeDetector(
       Driver, llvm::Triple(Driver.getTargetTriple()), ParsedArgs);
@@ -243,10 +244,13 @@ unsigned int GetLinesNumber(clang::tooling::RefactoringTool &Tool,
                             UnifiedPath Path) {
   // Set up Rewriter and to get source manager.
   LangOptions DefaultLangOptions;
-  IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts = new DiagnosticOptions();
-  TextDiagnosticPrinter DiagnosticPrinter(llvm::errs(), &*DiagOpts);
+  auto DiagOpts = std::make_shared<DiagnosticOptions>();
+
+  
+
+  TextDiagnosticPrinter DiagnosticPrinter(llvm::errs(), *DiagOpts);
   DiagnosticsEngine Diagnostics(
-      IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), &*DiagOpts,
+      IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), *DiagOpts,
       &DiagnosticPrinter, false);
   SourceManager Sources(Diagnostics, Tool.getFiles());
   Rewriter Rewrite(Sources, DefaultLangOptions);
@@ -679,9 +683,9 @@ int showAPIMapping(StringRef SrcAPI, StringRef Option, RefactoringTool &Tool,
                    ReplTy &ReplSYCL) {
   llvm::outs() << "CUDA API:" << llvm::raw_ostream::GREEN << SrcAPI
                << llvm::raw_ostream::RESET;
+  auto DiagOpts = std::make_shared<DiagnosticOptions>();
   DiagnosticsEngine Diagnostics(
-      IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()),
-      IntrusiveRefCntPtr<DiagnosticOptions>(new DiagnosticOptions()));
+      IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), *DiagOpts);
   SourceManager Sources(Diagnostics, Tool.getFiles());
   LangOptions DefaultLangOptions;
   Rewriter Rewrite(Sources, DefaultLangOptions);
diff --git a/clang/lib/DPCT/Diagnostics/Diagnostics.h b/clang/lib/DPCT/Diagnostics/Diagnostics.h
index 18c51833bf8f..aa9837646085 100644
--- a/clang/lib/DPCT/Diagnostics/Diagnostics.h
+++ b/clang/lib/DPCT/Diagnostics/Diagnostics.h
@@ -353,11 +353,12 @@ public:
 
 private:
   SourceManagerForWarning() {
-    DiagOpts = new DiagnosticOptions();
+
+    DiagOpts = std::make_shared<DiagnosticOptions>();
     DiagOpts->ShowColors = DpctGlobalInfo::getInstance().getColorOption();
-    DiagnosticPrinter = new TextDiagnosticPrinter(llvm::errs(), &*DiagOpts);
+    DiagnosticPrinter = new TextDiagnosticPrinter(llvm::errs(), *DiagOpts);
     Diagnostics = new DiagnosticsEngine(
-        IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), &*DiagOpts,
+        IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), *DiagOpts,
         DiagnosticPrinter, false);
 
     FSO.WorkingDir = ".";
@@ -367,7 +368,7 @@ private:
   }
 
   LangOptions DefaultLangOptions;
-  IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts;
+  std::shared_ptr<DiagnosticOptions> DiagOpts;
   TextDiagnosticPrinter *DiagnosticPrinter;
   DiagnosticsEngine *Diagnostics;
   FileSystemOptions FSO;
diff --git a/clang/lib/DPCT/FileGenerator/GenFiles.cpp b/clang/lib/DPCT/FileGenerator/GenFiles.cpp
index ab6b60d98ce2..737d81999d0e 100644
--- a/clang/lib/DPCT/FileGenerator/GenFiles.cpp
+++ b/clang/lib/DPCT/FileGenerator/GenFiles.cpp
@@ -111,12 +111,11 @@ static bool formatFile(const clang::tooling::UnifiedPath &FileName,
   // the first input (the original output of dpct without format), then the
   // result is wrong.
   clang::LangOptions DefaultLangOptions;
-  IntrusiveRefCntPtr<clang::DiagnosticOptions> DiagOpts =
-      new clang::DiagnosticOptions();
-  clang::TextDiagnosticPrinter DiagnosticPrinter(llvm::errs(), &*DiagOpts);
+  auto *DO = new DiagnosticOptions;
+  clang::TextDiagnosticPrinter DiagnosticPrinter(llvm::errs(), *DO);
   clang::DiagnosticsEngine Diagnostics(
       IntrusiveRefCntPtr<clang::DiagnosticIDs>(new clang::DiagnosticIDs()),
-      &*DiagOpts, &DiagnosticPrinter, false);
+      *DO, &DiagnosticPrinter, false);
 
   clang::FileSystemOptions FSO;
   FSO.WorkingDir = ".";
@@ -910,10 +909,10 @@ int saveNewFiles(clang::tooling::RefactoringTool &Tool,
   volatile ProcessStatus status = MigrationSucceeded;
   // Set up Rewriter.
   LangOptions DefaultLangOptions;
-  IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts = new DiagnosticOptions();
-  TextDiagnosticPrinter DiagnosticPrinter(llvm::errs(), &*DiagOpts);
+    auto DiagOpts = std::make_shared<DiagnosticOptions>();
+  TextDiagnosticPrinter DiagnosticPrinter(llvm::errs(), *DiagOpts);
   DiagnosticsEngine Diagnostics(
-      IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), &*DiagOpts,
+      IntrusiveRefCntPtr<DiagnosticIDs>(new DiagnosticIDs()), *DiagOpts,
       &DiagnosticPrinter, false);
   SourceManager Sources(Diagnostics, Tool.getFiles());
   Rewriter Rewrite(Sources, DefaultLangOptions);
diff --git a/clang/lib/DPCT/MigrateScript/GenMakefile.cpp b/clang/lib/DPCT/MigrateScript/GenMakefile.cpp
index b7d2e7f8b4ef..cfcd445d8c16 100644
--- a/clang/lib/DPCT/MigrateScript/GenMakefile.cpp
+++ b/clang/lib/DPCT/MigrateScript/GenMakefile.cpp
@@ -232,7 +232,8 @@ static void getCompileInfo(
 
         // Create minimalist CudaInstallationDetector to call the member
         // function validateCudaHeaderDirectory()
-        DiagnosticsEngine E(nullptr, nullptr, nullptr, false);
+        auto DiagOpts = std::make_shared<DiagnosticOptions>();
+        DiagnosticsEngine E(nullptr, *DiagOpts, nullptr, false);
         clang::driver::Driver Driver("", llvm::sys::getDefaultTargetTriple(),
                                      E);
         clang::driver::CudaInstallationDetector CudaIncludeDetector(
diff --git a/clang/lib/DPCT/MigrationAction.cpp b/clang/lib/DPCT/MigrationAction.cpp
index 64503e199740..fd3e4bb00cbe 100644
--- a/clang/lib/DPCT/MigrationAction.cpp
+++ b/clang/lib/DPCT/MigrationAction.cpp
@@ -181,22 +181,24 @@ std::shared_ptr<TranslationUnitInfo> DpctToolAction::createTranslationUnitInfoIm
     std::shared_ptr<CompilerInvocation> Invocation, bool &Success) {
   Invocation->getDiagnosticOpts().IgnoreWarnings = true;
   auto DiagConsumer = new TextDiagnosticPrinter(
-      DiagnosticStream, &Invocation->getDiagnosticOpts());
+      DiagnosticStream, Invocation->getDiagnosticOpts());
   auto Info = std::make_shared<TranslationUnitInfo>();
   FileSystemOptions FO;
   FileManager FileMgr(FO);
   auto Diags = CompilerInstance::createDiagnostics(
-      FileMgr.getVirtualFileSystem(), &Invocation->getDiagnosticOpts(),
+      FileMgr.getVirtualFileSystem(), Invocation->getDiagnosticOpts(),
       DiagConsumer,
       /*ShouldOwnClient=*/false, &Invocation->getCodeGenOpts());
   DpctGlobalInfo::setColorOption(Invocation->getDiagnosticOpts().ShowColors);
-  Info->AST = ASTUnit::create(Invocation, Diags, CaptureDiagsKind::None, false);
+  auto DiagOpts = std::make_shared<DiagnosticOptions>();
+  Info->AST = ASTUnit::create(Invocation, DiagOpts, Diags,
+                              CaptureDiagsKind::None, false);
   // Use the FileSystem passed by RefactoringTool.
   Info->AST->getFileManager().setVirtualFileSystem(FS);
   DpctFrontEndAction FEAction(Info.get());
   auto Ret = ASTUnit::LoadFromCompilerInvocationAction(
-      Invocation, std::make_shared<PCHContainerOperations>(), Diags, &FEAction,
-      Info->AST.get());
+      Invocation, std::make_shared<PCHContainerOperations>(), DiagOpts, Diags,
+      &FEAction, Info->AST.get());
   Success = !DiagConsumer->getNumErrors();
   if (Ret && (bool)&Info->AST->getASTContext())
     return Info;
diff --git a/clang/lib/DPCT/RuleInfra/ExprAnalysis.cpp b/clang/lib/DPCT/RuleInfra/ExprAnalysis.cpp
index da22c74d8c56..e1bbeade26b6 100644
--- a/clang/lib/DPCT/RuleInfra/ExprAnalysis.cpp
+++ b/clang/lib/DPCT/RuleInfra/ExprAnalysis.cpp
@@ -1006,7 +1006,6 @@ void ExprAnalysis::analyzeExpr(const CallExpr *CE) {
 
 void ExprAnalysis::analyzeExpr(const CXXMemberCallExpr *CMCE) {
   auto PP = DpctGlobalInfo::getContext().getPrintingPolicy();
-  PP.PrintCanonicalTypes = true;
   const auto ME = dyn_cast<MemberExpr>(CMCE->getCallee());
   if (!ME)
     return;
@@ -1177,7 +1176,6 @@ void ExprAnalysis::analyzeType(TypeLoc TL, const Expr *CSCE,
     TyName.clear();
     auto &TSTL = TYPELOC_CAST(TemplateSpecializationTypeLoc);
     auto PP = Context.getPrintingPolicy();
-    PP.PrintCanonicalTypes = 1;
     PrintFullTemplateName(OS, PP, TSTL.getTypePtr()->getTemplateName());
     if (!TypeLocRewriterFactoryBase::TypeLocRewriterMap)
       return;
diff --git a/clang/lib/DPCT/RulesAsm/AsmMigration.cpp b/clang/lib/DPCT/RulesAsm/AsmMigration.cpp
index 9baf21596145..898a96018939 100644
--- a/clang/lib/DPCT/RulesAsm/AsmMigration.cpp
+++ b/clang/lib/DPCT/RulesAsm/AsmMigration.cpp
@@ -1171,7 +1171,7 @@ protected:
 
       auto CommonStr = llvm::Twine(Str)
                            .concat("\"")
-                           .concat(GAS->getAsmString()->getString())
+                           .concat(GAS->getAsmString())
                            .concat("\"")
                            .str();
 
@@ -1240,7 +1240,7 @@ protected:
 
     if (DpctGlobalInfo::useSYCLCompat()) {
       report(Diagnostics::UNSUPPORT_SYCLCOMPAT, /*UseTextBegin=*/true,
-             GAS->getAsmString()->getString());
+             GAS->getAsmString());
       cutOffMigration();
       return SYCLGenSuccess();
     }
@@ -2431,7 +2431,7 @@ protected:
   bool HandleVset(const InlineAsmInstruction *I, StringRef Fn) {
     if (DpctGlobalInfo::useSYCLCompat()) {
       report(Diagnostics::UNSUPPORT_SYCLCOMPAT, /*UseTextBegin=*/true,
-             GAS->getAsmString()->getString());
+             GAS->getAsmString());
       cutOffMigration();
       return SYCLGenSuccess();
     }
@@ -3434,7 +3434,7 @@ protected:
   bool HandleCopyWait(const InlineAsmInstruction *Inst) {
     auto CommonStr = llvm::Twine("")
                          .concat("\"")
-                         .concat(GAS->getAsmString()->getString())
+                         .concat(GAS->getAsmString())
                          .concat("\"")
                          .str();
 
@@ -3467,7 +3467,7 @@ protected:
 
     auto CommonStr = llvm::Twine("")
                          .concat("\"")
-                         .concat(GAS->getAsmString()->getString())
+                         .concat(GAS->getAsmString())
                          .concat("\"")
                          .str();
 
@@ -3694,7 +3694,7 @@ void AsmRule::doMigrateInternel(const GCCAsmStmt *GAS) {
   std::string S = GAS->generateAsmString(C);
   InlineAsmContext Context;
   llvm::SourceMgr Mgr;
-  std::string Buffer = GAS->getAsmString()->getString().str();
+  std::string Buffer = GAS->getAsmString();
   Mgr.AddNewSourceBuffer(llvm::MemoryBuffer::getMemBuffer(Buffer),
                          llvm::SMLoc());
   SYCLIdentiferHandler Handle;
diff --git a/clang/lib/DPCT/RulesLang/Math/CallExprRewriterMath.cpp b/clang/lib/DPCT/RulesLang/Math/CallExprRewriterMath.cpp
index 0444ecc0b5b0..ff3349fa6159 100644
--- a/clang/lib/DPCT/RulesLang/Math/CallExprRewriterMath.cpp
+++ b/clang/lib/DPCT/RulesLang/Math/CallExprRewriterMath.cpp
@@ -468,7 +468,6 @@ std::optional<std::string> MathSimulatedRewriter::rewrite() {
     }
     auto &Context = dpct::DpctGlobalInfo::getContext();
     auto PP = Context.getPrintingPolicy();
-    PP.PrintCanonicalTypes = 1;
     auto Arg0 = Call->getArg(0);
     auto Arg1 = Call->getArg(1);
 
diff --git a/clang/lib/DPCT/RulesLangLib/CallExprRewriterThrust.cpp b/clang/lib/DPCT/RulesLangLib/CallExprRewriterThrust.cpp
index f0e8a7e424dc..7e295f2756b2 100644
--- a/clang/lib/DPCT/RulesLangLib/CallExprRewriterThrust.cpp
+++ b/clang/lib/DPCT/RulesLangLib/CallExprRewriterThrust.cpp
@@ -129,7 +129,6 @@ inline std::function<std::string(const CallExpr *)>
 makeMappedThrustPolicyEnum(unsigned Idx) {
   auto getBaseType = [](QualType QT) -> std::string {
     auto PP = DpctGlobalInfo::getContext().getPrintingPolicy();
-    PP.PrintCanonicalTypes = true;
     return QT.getUnqualifiedType().getAsString(PP);
   };
   auto getMehtodName = [](const ValueDecl *VD) -> std::string {
diff --git a/clang/lib/DPCT/RulesLangLib/ThrustAPIMigration.cpp b/clang/lib/DPCT/RulesLangLib/ThrustAPIMigration.cpp
index efe87c2a06c5..b5e85b1f06b3 100644
--- a/clang/lib/DPCT/RulesLangLib/ThrustAPIMigration.cpp
+++ b/clang/lib/DPCT/RulesLangLib/ThrustAPIMigration.cpp
@@ -108,7 +108,6 @@ void ThrustAPIRule::thrustFuncMigration(const MatchFinder::MatchResult &Result,
   const CXXMemberCallExpr *CMCE = dyn_cast_or_null<CXXMemberCallExpr>(CE);
   if (CMCE) {
     auto PP = DpctGlobalInfo::getContext().getPrintingPolicy();
-    PP.PrintCanonicalTypes = true;
     auto BaseType = CMCE->getObjectType().getUnqualifiedType().getAsString(PP);
     StringRef BaseTypeRef(BaseType);
     if (BaseTypeRef.starts_with("thrust::cuda_cub::execute_on_stream_base<") &&
diff --git a/clang/lib/DPCT/Utility.cpp b/clang/lib/DPCT/Utility.cpp
index e9b2e30b6e17..b08c6d87d978 100644
--- a/clang/lib/DPCT/Utility.cpp
+++ b/clang/lib/DPCT/Utility.cpp
@@ -22,6 +22,7 @@
 #include "clang/AST/DeclBase.h"
 #include "clang/AST/Expr.h"
 #include "clang/AST/ExprCXX.h"
+#include "clang/Basic/OperatorKinds.h"
 #include "clang/Basic/SourceLocation.h"
 #include "clang/Basic/SourceManager.h"
 #include "clang/Basic/TokenKinds.h"
@@ -4794,10 +4795,11 @@ void PrintFullTemplateName(raw_ostream &OS, const PrintingPolicy &Policy, Templa
   } else if (DependentTemplateName *DTN = Name.getAsDependentTemplateName()) {
     OS << "template ";
 
-    if (DTN->isIdentifier())
-      OS << DTN->getIdentifier()->getName();
+    
+    if (DTN->getName().getOperator() == OO_None)
+      OS << DTN->getName().getIdentifier()->getName();
     else
-      OS << "operator " << getOperatorSpelling(DTN->getOperator());
+      OS << "operator " << getOperatorSpelling(DTN->getName().getOperator());
   } else if (SubstTemplateTemplateParmStorage *subst
                = Name.getAsSubstTemplateTemplateParm()) {
     PrintFullTemplateName(OS, Policy, subst->getReplacement());
